### PR TITLE
[stdlib] Unicode 9 and Thread Local Storage

### PR DESCRIPTION
**Pull request by milseman**
TL;DR Add support for Unicode 9 grapheme breaking (i.e. emoji) by using ICU. Work around ICU efficiency issues with thread-local storage.

---

Introduce shims for using UBreakIterators from ICU. Also introduce
shims for using thread local storage via pthreads.
    
We will be relying on ICU and UBreakIterators for grapheme
breaking. But, UBreakIterators are very expensive to create,
especially for the way we do grapheme breaking, which is relatively
stateless. Thus, we will stash one or more into thread local storage
and reset it as needed.
    
---

Introduces a _ThreadLocalStorage struct to hold thread-local, but
global resources. Set it up to host a UBreakIterator and a cache key
for resetting text.
    
UBreakIterators are extremely expensive to create on the fly, so we
store one for each thread. UBreakIterators are also expensive to bind
to new text, so we cache the text it's currently bound to in order to
help avoid it.
    
The struct can be expanded with more functionality in the future, but
the standard library should only ever use a single key, and thus
everything should go on this struct. The _ThreadLocalStorage struct is
not meant to be copyable, creatable (by anyone else except the
once-per-thread initialize routine), and should accessed through the
pointers it provides.
    
Future immediate directions could include cashing multiple
UBreakIterators (e.g. avoid a text reset for mutual character
iteration patterns, etc).
    
Test added in test/stdlib/ThreadLocalStorage.swift.

---

Use UBreakIterators to perform grapheme breaking. This gives Unicode 9
grapheme breaking (e.g. family emoji) and provides a means to upgrade
to future versions. It also serves as a model for how to serve up
other advanced functionality in ICU to users.
    
This has tricky performance implications. Some things are faster and a
number of cases are slower. But, careful use of ICU can help mitigate
and amortize these costs. In conjunction with more early detection of
fast paths, overall grapheme breaking for the average user should be
much faster than in Swift 3.
    
NOTE: This is incomplete. It currently falls back on the legacy tries
for some bridged strings. There are many potential directions for a
general solution, but for now we'll be interatively adding support for
more and more special cases.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test 

Not just for kicks, I'm genuinely interested

File: stdlib/public/core/CMakeLists.txt [hunk 1]
```diff
@@ -174,6 +176,7 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
   list(APPEND swift_core_link_flags "-all_load")
   list(APPEND swift_core_framework_depends Foundation)
   list(APPEND swift_core_framework_depends CoreFoundation)
+  list(APPEND swift_core_private_link_libraries -licucore)
```
Comment (milseman) [inline stdlib/public/core/CMakeLists.txt]: As @compnerd mentioned, we can remove the "-l" part

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
I think those benchmarks are showing changes on master from other things that went in since, like the new utf8 decoding and the change to remove <, not this change. Needs rebasing and rebenchmarking.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
Rewrote history a little to squash some things, and only pull in the tiny bits of ICU that we use.

**Comment by milseman** [comment]
Added in more static protections for cross-platform differences and a cache coherency fix.

**Comment by milseman** [comment]
@swift-ci please smoke test Linux platform

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
I extracted some of those smoke benchmarks, and I'm seeing us being way faster with this PR compared to a couple weeks ago (before a lot of my other optimizations). I'm measuring that we are:
StringWithCString: 5x faster
CStringLongAscii: 2.5x faster
CStringLongNonAscii: 2x faster

than a couple weeks ago.

**Comment by milseman** [comment]
```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/lib/ClangImporter/ClangImporter.cpp:2647:41: error: too few arguments to function call, single argument 'Diags' was not specified
  return Impl.Invocation->getModuleHash();
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```

looks like the builder's broken, as other PRs are hitting this. Will try again soon.

**Comment by milseman** [comment]
@swift-ci please smoke test Linux platform

**Comment by CodaFi** [comment]
Just for bookkeeping: This supersedes #3125, right?

**Comment by milseman** [comment]
@CodaFi when it lands, yes

**Comment by milseman** [comment]
Sweet, that looks better. I'll look into those.

**Comment by milseman** [comment]
Linux platform defines a different pthread_key_t, as suspected. Will fix.

**Comment by milseman** [comment]
@swift-ci please smoke test Linux Platform

**Comment by milseman** [comment]
SuperChars is a legit slowdown. Grapheme breaking performance is sometimes faster and sometimes slower than the tries if we can't hit our fast paths. 

Some of the delta is the same inefficient Character construction from a (presumably) valid CharcterView.Index, but not most.

@airspeedswift Any ideas? Are you ok with this?

**Comment by airspeedswift** [comment]
I think the regression in SuperChars is fine, that benchmark isn't very representative of real usage.

**Comment by milseman** [comment]
@swift-ci please smoke test Linux Platform

**Comment by milseman** [comment]
@swift-ci please test OS X Platform

**Comment by milseman** [comment]
@swift-ci please smoke test Linux Platform

**Comment by milseman** [comment]
@swift-ci please smoke test Linux Platform

**Comment by milseman** [comment]
@swift-ci please smoke test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
Just to be extra cautious, I'm going to wait for that OS X test to finish...
